### PR TITLE
fix: Race conditions with multiple downloads on the same layer

### PR DESCRIPTION
**Pull request by dkphm**
#### Why is this change necessary?
Customer reports that when the SAM template contains multiple functions, some shared the same lambda layer, `local start-api` throw the exception

```
2024-08-27 09:43:43,083 | Lambda functions containers initialization failed because of [Errno 2] No such file or directory: '/Users/aguro/.aws-sam/layers-pkg/NFTUserLayer-3-ff9c237555.zip'
```

This is because there was a race condition when multiple layer_downloaders were running to build multiple functions concurrently. Some could finish downloading the file first and remove the .zip file, causing the other one failed to unzip.

#### How does it address the issue?
The fix is to ask each downloader to download into separate zip file (by adding the thread id to the file name), but the extracted folder remains the same.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [X] Write/update unit tests
- [X] Write/update integration tests
- [X] Write/update functional tests if needed
- [X] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/local/layers/layer_downloader.py [hunk 1]
```diff
@@ -106,7 +107,8 @@ def download(self, layer: LayerVersion, force=False) -> LayerVersion:
             LOG.info("%s is already cached. Skipping download", layer.arn)
             return layer
 
-        layer_zip_path = layer.codeuri + ".zip"
+        current_thread_id = str(threading.get_ident())
+        layer_zip_path = f"{layer.codeuri}_{current_thread_id}.zip"
```
Comment (mndeveci) [inline samcli/local/layers/layer_downloader.py:111]: If we are looking for getting a unique value, why we are not using `uuid`?

Comment (dkphm) [reply samcli/local/layers/layer_downloader.py:111]: Oh I was thinking about one file per thread, but yeah using `uuid` is also good for this case.

**Review by mndeveci — APPROVED** [review]

File: samcli/local/layers/layer_downloader.py [hunk 2]
```diff
@@ -106,7 +107,7 @@ def download(self, layer: LayerVersion, force=False) -> LayerVersion:
             LOG.info("%s is already cached. Skipping download", layer.arn)
             return layer
 
-        layer_zip_path = layer.codeuri + ".zip"
+        layer_zip_path = f"{layer.codeuri}_{uuid.uuid4().hex}.zip"
         layer_zip_uri = self._fetch_layer_uri(layer)
         unzip_from_uri(
```
Comment (lucashuy) [inline samcli/local/layers/layer_downloader.py:112]: quick question, but will we come across any issues when the Layer is unpacked into the same directory across different threads?

Comment (dkphm) [reply samcli/local/layers/layer_downloader.py:112]: Oh it should be fine. This is what we want :)

**Review by lucashuy — APPROVED** [review]

**Pull request merged**